tests/test_orient_symmetric.py

```diff
@@ -54,19 +224,113 @@ def testEstimateThirdRows(self):
 
         # Estimate third rows from outer products.
         # Due to factorization of V, these might be negated third rows.
-        vis = self.cl_class._estimate_third_rows(vijs, viis)
+        vis = self.cl_classes[3]._estimate_third_rows(vijs, viis)
 
         # Check if all-close up to difference of sign
         ground_truth = np.sign(gt_vis[0, 0]) * gt_vis
         estimate = np.sign(vis[0, 0]) * vis
         self.assertTrue(np.allclose(ground_truth, estimate))
 
+    @parameterized.expand([(order,), (order + 1,)])
+    def testSelfCommonLines(self, order):
+        n_img = self.n_img
+        n_theta = self.n_theta
+        src = self.srcs[order]
+        cl_symm = self.cl_classes[order]
+
+        # Initialize common-lines class and compute self-common-lines matrix.
+        scl, _, _ = cl_symm._self_clmatrix_c3_c4()
+
+        # Compute ground truth self-common-lines matrix.
+        rots = src.rots
+        scl_gt = self.buildSelfCommonLinesMatrix(rots, order)
+
+        # Get angle difference between scl_gt and scl.
+        scl_diff1 = scl_gt - scl
+        scl_diff2 = scl_gt - np.flip(scl, 1)  # Order of indices might be switched.
+        scl_diff1_angle = scl_diff1 * 2 * pi / n_theta
+        scl_diff2_angle = scl_diff2 * 2 * pi / n_theta
+
+        # cosine is invariant to 2pi, and abs is invariant to +-pi due to J-conjugation.
+        # We take the mean deviation wrt to the two lines in each image.
+        scl_diff1_angle_mean = np.mean(np.arccos(abs(np.cos(scl_diff1_angle))), axis=1)
+        scl_diff2_angle_mean = np.mean(np.arccos(abs(np.cos(scl_diff2_angle))), axis=1)
+
+        scl_diff_angle_mean = np.vstack((scl_diff1_angle_mean, scl_diff2_angle_mean))
+        scl_idx = np.argmin(scl_diff_angle_mean, axis=0)
+        min_mean_angle_diff = scl_idx.choose(scl_diff_angle_mean)
+
+        # Assert scl detection rate is greater than 90% with 1 degree tolerance for order=3,
+        # and 3 degree tolerance for order=4.
+        if order == 3:
+            angle_tol_err = 1 * pi / 180
+        else:
+            angle_tol_err = 3 * pi / 180
+
+        detection_rate = np.count_nonzero(min_mean_angle_diff < angle_tol_err) / n_img
+        self.assertTrue(detection_rate > 0.90)
+
+    @parameterized.expand([(order,), (order + 1,)])
+    def testCommonLines(self, order):
+        n_img = self.n_img
+        src = self.srcs[order]
+        cl_symm = self.cl_classes[order]
+
+        # Build common-lines matrix.
+        cl_symm.build_clmatrix()
+        cl = cl_symm.clmatrix
+
+        # Compare common-line indices with ground truth angles.
+        rots = src.rots  # ground truth rotations
+        rots_symm = self.buildCyclicRotations(order)
+        pairs = all_pairs(n_img)
+        within_1_degree = 0
+        within_5_degrees = 0
+        for (i, j) in pairs:
+            a_ij_s = np.zeros(order)
+            a_ji_s = np.zeros(order)
+            cl_ij = cl[i, j] % 180
```

Comment:
Avoid hardcoding. Use `n_theta // 2`.

Reply:
This is actually meant to be `180` and not `n_theta // 2`. We are taking the index from the half of the common line that lies in [0, 180).

Reply:
Why? What meaning does 180 have in our indexing system?

Reply:
You're correct here. Faulty thinking on my part. Thanks. changed.